docs/api/ref/api-v3.yml

```diff
@@ -247,6 +247,11 @@ paths:
                         description: Array of sorted strings suggestions in the language requested in the "lc" field.
                         items:
                           type: string
+                      matched_synonyms:
```

Comment:
from what I undertand seeing tests, you did not only change api-v3 but also api-v2 (in api.yml). shan't you change it there also ?

Reply:
This comment might not be relevant if we go for a suggestion api v2.

Reply:
I am not sure where I should separate my changes. Do I copy [this part](https://github.com/openfoodfacts/openfoodfacts-server/blob/417a4c0de43b234b07f608400fdbf12aaf38c25d/lib/ProductOpener/API.pm#L405) with a new name with the new function exposed only there? If so, do I need to copy [this function](https://github.com/openfoodfacts/openfoodfacts-server/blob/086a0a0a926589aecf14a1f33e12785a1c1270c5/lib/ProductOpener/APITaxonomySuggestions.pm#L67) as well? Or do the changes need to be separated at the level of inner functions, like from [here](https://github.com/openfoodfacts/openfoodfacts-server/blob/086a0a0a926589aecf14a1f33e12785a1c1270c5/lib/ProductOpener/TaxonomySuggestions.pm#L317)?

I am also not sure how I should change `api.yml`, since [it appears mostly empty](https://github.com/openfoodfacts/openfoodfacts-server/blob/086a0a0a926589aecf14a1f33e12785a1c1270c5/docs/api/ref/api.yml#L277).

Reply:
@Naruyoko in the code you show, condition should be enough to sort out the cases.

But we could add a parameter to `get_taxonomy_suggestions` and have a `cgi/suggest_v2.pl` alike [suggest.pl](https://github.com/openfoodfacts/openfoodfacts-server/blob/417a4c0de43b234b07f608400fdbf12aaf38c25d/cgi/suggest.pl)

Reply:
@Naruyoko, do not hesitate to ping me on slack if it's not clear enough.

Reply:
After discussion:

* do not change suggest.pl output
* we can change api/v3/suggest output conditioned by a specific parameter (with_synonyms?)
* we will ask the mobile app to use this new parameter so that we can remove the old output for api/v3/suggest (easier to match to OpenAPI)

Reply:
The output of suggest.pl should be reverted now. It wasn't clear to me that I should expose the feature through a new CGI.

I think we should discuss the usage and the returned structure for the new v3. Currently,

- The synonym information is controlled by the new `get_synonyms` flag being set to `1`.
- The flag adds the list of synonyms as `matched_synonyms` to the data (See [here](https://github.com/openfoodfacts/openfoodfacts-server/pull/9395#issuecomment-1839231773)). They are in the same order as the list of tags.

I think of two considerations:
- What would be an appropriate name for this flag (you put `with_synonyms`)? Would it even make sense to switch the behavior with a such flag?
- Is there a better format? Maybe they should be transposed so it returns an array of tag-synonym pairs (as it is internally)?

Reply:
> This just occurred to me: is the "v3" API the in-dev/evolving version?

Exactly :-)

Reply:
> Is there a better format? Maybe they should be transposed so it returns an array of tag-synonym pairs (as it is internally)?

I thinks it's better to use dict in an API, as you dead, it's cleaner and more "self documented".

Reply:
By using dict, do you mean like this (which was what I intended by "traspose"):

```json
{"errors":[],"suggestions":[{"matched_synonym":"Orge","tag":"Gluten"},{"matched_synonym":"Œufs","tag":"Œufs"},{"matched_synonym":"Arachis hypogaea","tag":"Arachides"},{"matched_synonym":"Langoustine","tag":"Crustacés"},{"matched_synonym":"Fruits à coque dure","tag":"Fruits à coque"},{"matched_synonym":"Beurre concentré","tag":"Lait"},{"matched_synonym":"Pétoncles","tag":"Mollusques"},{"matched_synonym":"Graines de moutarde","tag":"Moutarde"},{"matched_synonym":"Rouget","tag":"Poisson"},{"matched_synonym":"Tofu","tag":"Soja"}],"warnings":[]}
```

Or like this:

```json
{"errors":[],"suggestions":{"Gluten":"Orge","Œufs":"Œufs","Arachides":"Arachis hypogaea","Crustacés":"Langoustine","Fruits à coque":"Fruits à coque dure","Lait":"Beurre concentré","Mollusques":"Pétoncles","Moutarde":"Graines de moutarde","Poisson":"Rouget","Soja":"Tofu"},"warnings":[]}
```

Or like this:

```json
{"errors":[],"matched_synonyms":{"Gluten":"Orge","Œufs":"Œufs","Arachides":"Arachis hypogaea","Crustacés":"Langoustine","Fruits à coque":"Fruits à coque dure","Lait":"Beurre concentré","Mollusques":"Pétoncles","Moutarde":"Graines de moutarde","Poisson":"Rouget","Soja":"Tofu"},"status":"success","suggestions":["Gluten","Œufs","Arachides","Crustacés","Fruits à coque","Lait","Mollusques","Moutarde","Poisson","Soja"],"warnings":[]}
```

Reply:
I like your third version, because it keeps things simple as you can just retrieve the simple suggestion list, and it's easy to find back matched synonym.

Just one point: shan't it be `"matched_synonyms":{"Gluten":["Orge"],…}` as there might be more than one matched synonym (or don't we care because we just return one ? Which is still really ok for me, as it covers the needs). 

Also is it useful to put entries like "Œufs": "Œufs" inside matched_synonyms or can the API user deduce from the fact that Œufs is not in the synonyms dict that it means it was directly matched ?

Reply:
I agree with using the third version.

The matched synonym is currently the first matched one of the best match (with "start">"inside">"fuzzy"). I don't see how much useful it is to return all possible matches.

I think it is better to keep `"a":"a"` entries for consistency.

Reply:
I pushed a change with the new format. I will have to update the API docs and create new tests.

Reply:
Yes, I read your changes, seems perfect :muscle: 

So I wait for the other changes. Don't hesitate to ping me on slack when you are ready @Naruyoko :-)